### PR TITLE
(RFC) logging: app-layer and proto detect layer events

### PR DESCRIPTION
**Pull request by jlucovsky**
_Request for comments_

This PR includes changes for 2 issues:
- [2941](https://redmine.openinfosecfoundation.org/issues/2941)
- [2942](https://redmine.openinfosecfoundation.org/issues/2942)

Continuation of #3867 

App-layer events and protocol detection events are now logged to the anomaly log with changes in this pr.

App-layer events are logged by adding a "wild card" application layer logger. This logger is invoked while traversing the TX loggers and is invoked for all TX, regardless of the `alproto`; hence, the term "wild-card". This logger may be invoked multiple times for the same TX; a mechanism to track which events have been logged previously is employed to avoid logging an individual event twice.

Protocol detection events are logged when `app_layer_events` has an event posted. Parser events are logged when posted to the (flow) `alparser` area.

App layer parsers have been amended to include a function to translate an application layer parser specific event id into a display string.

### TODO
- Extend each app-layer-parser to support the new `GetEventInfoById` function that translates an app layer parser event value into a display string.
- Documentation updates.

#### Observations
Each transaction has an event set; for a given transaction, there are sometimes multiple, back-to-back events recorded that are the same. 

Would like some feedback on the approach as well as the content/key naming of generated log records (below).

### Protocol Detection Event
```
{
  "timestamp": "2014-09-09T18:29:28.179967-0700",
  "flow_id": 1831982488608705,
  "pcap_cnt": 1077,
  "event_type": "anomaly",
  "src_ip": "172.16.165.134",
  "src_port": 25,
  "dest_ip": "172.16.165.133",
  "dest_port": 49200,
  "proto": "TCP",
  "anomaly": {
    "alproto": "smtp",
    "type": "packet",
    "event": "APPLAYER_DETECT_PROTOCOL_ONLY_ONE_DIRECTION",
    "event_no": "1 (of 1)",
    "layer": "app_layer"
  }
}
```

### App layer events
```
{
  "timestamp": "2014-10-06T16:10:03.482379-0700",
  "flow_id": 1831982488608705,
  "event_type": "anomaly",
  "src_ip": "172.16.165.134",
  "src_port": 25,
  "dest_ip": "172.16.165.133",
  "dest_port": 49200,
  "proto": "TCP",
  "tx_id": 0,
  "anomaly": {
    "alproto": "smtp",
    "type": "transaction",
    "event": "INVALID_REPLY",
    "event_no": "1 (of 1)",
    "layer": "applayer_parser"
  }
}
{
  "timestamp": "2015-01-06T14:27:26.548474-0800",
  "flow_id": 833463658664645,
  "pcap_cnt": 636813,
  "event_type": "anomaly",
  "src_ip": "108.61.179.182",
  "src_port": 443,
  "dest_ip": "192.168.138.163",
  "dest_port": 49168,
  "proto": "TCP",
  "tx_id": 0,
  "anomaly": {
    "alproto": "http",
    "type": "transaction",
    "event": "REQUEST_AUTH_UNRECOGNIZED",
    "event_no": "1 (of 1)",
    "layer": "applayer_parser"
  }
}
{
  "timestamp": "2015-01-06T14:27:27.959384-0800",
  "flow_id": 833463658664645,
  "pcap_cnt": 636816,
  "event_type": "anomaly",
  "src_ip": "192.168.138.163",
  "src_port": 49168,
  "dest_ip": "108.61.179.182",
  "dest_port": 443,
  "proto": "TCP",
  "tx_id": 0,
  "anomaly": {
    "alproto": "http",
    "type": "transaction",
    "event": "GZIP_DECOMPRESSION_FAILED",
    "event_no": "2 (of 5)",
    "layer": "applayer_parser"
  }
}
{
  "timestamp": "2011-01-25T10:54:08.507043-0800",
  "flow_id": 1685566832748781,
  "pcap_cnt": 4232,
  "event_type": "anomaly",
  "src_ip": "194.165.188.79",
  "src_port": 443,
  "dest_ip": "172.16.255.1",
  "dest_port": 10627,
  "proto": "TCP",
  "tx_id": 0,
  "anomaly": {
    "alproto": "tls",
    "type": "transaction",
    "event": "INVALID_SSL_RECORD",
    "event_no": "4 (of 4)",
    "layer": "applayer_parser"
  }
}
```

### CODE REVIEW

File: src/output-json-anomaly.c [hunk 1]
```diff
@@ -115,25 +119,152 @@ static int AnomalyJson(ThreadVars *tv, JsonAnomalyLogThread *aft, const Packet *
...
+static int AnomalyAppLayerDecoderEventJson(JsonAnomalyLogThread *aft,
+                        const Packet *p, AppLayerDecoderEvents *decoder_events,
+                        bool is_applayer, const char *layer, uint64_t tx_id)
+{
+    const char *alprotoname = AppLayerGetProtoName(p->flow->alproto);
+
+    SCLogDebug("decoder_events %p event_count %d (last logged %d) %s",
+                decoder_events, decoder_events->cnt,
+                decoder_events->event_last_logged,
+                tx_id != ((uint64_t)-1) ? "tx" : "no-tx");
```
Comment (victorjulien) [inline src/output-json-anomaly.c]: this -1 use on an unsigned type is rather ugly. Why not use UINT64_MAX? Might also be nicer to use a define like 
```c
#define TX_ID_UNUSED UINT64_MAX
```
or something.

File: src/output-json-anomaly.c [hunk 2]
```diff
@@ -115,25 +119,152 @@ static int AnomalyJson(ThreadVars *tv, JsonAnomalyLogThread *aft, const Packet *
...
         /* anomaly */
-        json_object_set_new(js, "anomaly", ajs);
+        json_object_set_new(js, ANOMALY_EVENT_TYPE, ajs);
         OutputJSONBuffer(js, aft->file_ctx, &aft->json_buffer);
 
         json_object_clear(js);
         json_decref(js);
+
+        /* Current implementation assumes a single owner for this value */
+        ++decoder_events->event_last_logged;
```
Comment (victorjulien) [inline src/output-json-anomaly.c]: style: we use
```c
decoder_events->event_last_logged++
```

File: src/output-json-anomaly.c (hunk 1 above)
Comment (jlucovsky) [reply src/output-json-anomaly.c]: Will modify.

File: src/output-json-anomaly.c (hunk 2 above)
Comment (jlucovsky) [reply src/output-json-anomaly.c]: Will update.

**Comment by jlucovsky** [comment]
Continued in #3884

**Comment by victorjulien** [comment]
Btw I ran QA overnight and I'm seeing one serious failure:
```
==19698== Command: ./src/suricata -c suricata.yaml --disable-detection -v -l /tmp/ --set classification-file=classification.config --set reference-config-file=reference.config -k none -r /qa/sv/public-smb/public-wiresharkwiki-smb1-smbtorture-01/smbtorture.pcap
==19698== 
[19698] 22/5/2019 -- 20:13:53 - (suricata.c:1073) <Notice> (LogVersion) -- This is Suricata version 5.0.0-dev running in USER mode
[19698] 22/5/2019 -- 20:13:53 - (util-cpu.c:171) <Info> (UtilCpuPrintSummary) -- CPUs/cores online: 56
[19698] 22/5/2019 -- 20:13:56 - (util-logopenfile.c:476) <Info> (SCConfLogOpenGeneric) -- fast output device (regular) initialized: fast.log
[19698] 22/5/2019 -- 20:13:56 - (util-logopenfile.c:476) <Info> (SCConfLogOpenGeneric) -- eve-log output device (regular) initialized: eve.json
[19698] 22/5/2019 -- 20:13:56 - (util-logopenfile.c:476) <Info> (SCConfLogOpenGeneric) -- stats output device (regular) initialized: stats.log
[19698] 22/5/2019 -- 20:14:10 - (tm-threads.c:2157) <Notice> (TmThreadWaitOnThreadInit) -- all 57 packet processing threads, 4 management threads initialized, engine started.
[19699] 22/5/2019 -- 20:14:10 - (source-pcap-file.c:176) <Info> (ReceivePcapFileLoop) -- Starting file run for /qa/sv/public-smb/public-wiresharkwiki-smb1-smbtorture-01/smbtorture.pcap
[19699] 22/5/2019 -- 20:14:13 - (util-checksum.c:89) <Info> (ChecksumAutoModeCheck) -- No packets with invalid checksum, assuming checksum offloading is NOT used
[19699] 22/5/2019 -- 20:27:53 - (source-pcap-file-helper.c:149) <Info> (PcapFileDispatch) -- pcap file /qa/sv/public-smb/public-wiresharkwiki-smb1-smbtorture-01/smbtorture.pcap end of file reached (pcap err code 0)
[19698] 22/5/2019 -- 20:27:53 - (suricata.c:2843) <Notice> (SuricataMainLoop) -- Signal Received.  Stopping engine.
[19698] 22/5/2019 -- 20:28:55 - (tm-threads.c:1512) <Warning> (TmThreadDrainPacketThreads) -- [ERRCODE: SC_ERR_SHUTDOWN(188)] - unable to get all packet threads to process their packets in time
[19698] 22/5/2019 -- 20:30:02 - (tm-threads.c:1512) <Warning> (TmThreadDrainPacketThreads) -- [ERRCODE: SC_ERR_SHUTDOWN(188)] - unable to get all packet threads to process their packets in time
[19698] 22/5/2019 -- 20:31:03 - (tm-threads.c:1663) <Error> (TmThreadDisablePacketThreads) -- [ERRCODE: SC_ERR_FATAL(171)] - Engine unable to disable detect thread - "W#01". Killing engine
```
This might indicate we got stuck. The pcap is smb torture from https://wiki.wireshark.org/SampleCaptures

**Pull request closed without merging**